pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/TriggerSomaticResultSurveyRoute.java

```diff
@@ -0,0 +1,149 @@
+            long triggerId = buildTriggerComment(userId);
+            SomaticResultUpload triggerUpdatedSomaticUpload =
+                    service.updateSomaticResultTrigger(resultUpload.getSomaticDocumentId(), triggerId, realm);
+            SomaticResultTriggerActivityPayload payloadToSend = new SomaticResultTriggerActivityPayload(
+                    requestPayload.getParticipantId(),
+                    triggerId,
+                    triggerUpdatedSomaticUpload.getBucket(),
+                    triggerUpdatedSomaticUpload.getBlobPath());
+            return DDPRequestUtil.triggerFollowupSurvey(instance, payloadToSend, requestPayload.getSurveyName());
+        } else {
+            throw new DSMBadRequestException("Bad somatic document id.  Not triggering followup");
+        }
+    }
+
+    private boolean isEligibleToBeSent(SomaticResultUpload resultUpload) {
+        if (Boolean.FALSE.equals(resultUpload.getIsVirusFree())) {
+            throw new DSMBadRequestException(
+                    "The file selected is either not finished being scanned for viruses or was removed because one was found.");
+        }
+        if (resultUpload.getDeletedAt() > 0) {
+            throw new DSMBadRequestException("The file selected was already deleted and cannot be sent.");
+        }
+        return true;
+    }
+
+    private long buildTriggerComment(String userIdRequest) {
+        long currentTime = System.currentTimeMillis();
+        try {
+            return addTriggerCommentIntoDB(userIdRequest, currentTime);
+        } catch (Exception ex) {
+            throw new DsmInternalError("Error encountered adding trigger and comment, please contact a DSM developer.");
+        }
+    }
+
+    //this is modeled after the Query parameters used in TriggerSurveyRoute.java because we want this endpoint
+    //to be similar to that one, but add the support for the required documentId parameter that we need to look
+    //up the correct somatic file that was uploaded.
+    private SomaticResultTriggerRequestPayload getRequestPayload(QueryParamsMap queryParams, Request request) {
+        SomaticResultTriggerRequestPayload result = new SomaticResultTriggerRequestPayload();
+        if (queryParams.value("surveyName") != null) {
+            result.setSurveyName(queryParams.get("surveyName").value());
+        } else {
+            throw new DSMBadRequestException("No surveyName query param was sent");
+        }
+        if (!result.getSurveyName().equalsIgnoreCase("SOMATIC_RESULTS")) {
+            throw new DSMBadRequestException("This route only works with SOMATIC_RESULTS surveys.");
+        }
+        if (queryParams.value("surveyType") != null) {
+            result.setSurveyType(queryParams.get("surveyType").value());
+        } else {
+            throw new DSMBadRequestException("No surveyType query param was sent");
+        }
+        if (!result.getSurveyType().equalsIgnoreCase("REPEATING")) {
```

Comment:
Dumb question but why do we require the request to include a survey type, if only was in valid? Is there a larger pattern? Might this change in the future?

Reply:
It mimics the current endpoint and defines the rules that state this follow-up survey can be triggered multiple times.